torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,110 @@
+            Performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step`
+            using the gradients in the given :class:`DistributedDataParallel`
+            gradient bucket.
+            """
+            # Proceed as normal until the DDP buckets have been rebuilt
+            if not ddp._has_rebuilt_buckets:
+                return [bucket.get_tensor()]
+
+            bucket_index = bucket.get_index()
+            rank = zero.global_rank
+            overlap_info = zero._overlap_info
+            if overlap_info.status == _OverlapStatus.UNINITIALIZED:
+                overlap_info.status = _OverlapStatus.DDP_HAS_REBUILT_BUCKETS
+
+            bucket_params = bucket.get_model_params_for_bucket()
+            assert len(bucket_params) > 0, "Empty bucket"
+            rank_to_update = zero._ddp_bucket_index_to_rank(bucket_index)
+
+            # Once DDP buckets have been rebuilt but ZeRO has not been
+            # properly initialized yet, collect the information needed
+            if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
+                params_per_rank = overlap_info.params_per_rank
+                params_per_bucket = overlap_info.params_per_bucket
+                if rank_to_update == rank:
+                    overlap_info.offsets[bucket_index] = len(params_per_rank[rank_to_update])
+                params_per_rank[rank_to_update].extend(bucket_params)
+                params_per_bucket.append(bucket_params)
+
+                return [bucket.get_tensor()]
+
+            if rank_to_update == rank:
+                assert len(zero.optim.param_groups) == 1, \
+                    "Overlapping DDP with ZeRO only supports a single " \
+                    "parameter group"
+                # Construct the `gradients` input for the local optimizer step,
+                # which expects `None` in a list position to indicate that the
+                # corresponding parameter should not be updated
+                num_local_optim_params = len(zero.optim.param_groups[0]["params"])
+                gradients: List[Optional[torch.Tensor]] = \
+                    [None for _ in range(num_local_optim_params)]
+                assert bucket_index in overlap_info.offsets, \
+                    f"Bucket index {bucket_index} was not assigned to rank " \
+                    f"{rank}"
+                offset = overlap_info.offsets[bucket_index]
+                bucket_gradients = bucket.get_per_parameter_tensors()
+                for i, grad in enumerate(bucket_gradients):
+                    gradients[offset + i] = grad
+                zero._local_step(gradients)
+
+            device = bucket_params[0].device
+            device_index = zero._device_to_device_index[device]
+            assert bucket_index in zero._buckets[device_index][rank_to_update]
+            dist.broadcast(zero._buckets[device_index][rank_to_update][bucket_index], src=rank_to_update, async_op=True)
+
+            return [bucket.get_tensor()]
```

Comment:
Would I be correct if I assume that since this returned gradients won't be used anyway, we can return arbitrary Tensor values?

Reply:
I do not believe that assumption is correct. I think that we must still return the bucket's tensor since the value of the new `Future` returned by `then()` will be used in `Reducer::finalize_backward()`. See [here](https://github.com/pytorch/pytorch/blob/04bd9d7577ff855419adca1eb59b58d0f73a1a18/torch/csrc/distributed/c10d/reducer.cpp#L1449).

Reply:
Although maybe it shouldn't matter, since the returned gradients are basically just copied back into `param.grad` field, but param.grad field may be irrelevant if we are doing an explicit step on the gradients here.

Although I think you mentioned that before buckets are rebuilt user will still call `zero.step()`, in that case the correct gradient tensor indeed needs to be returned.

Reply:
> Although I think you mentioned that before buckets are rebuilt user will still call zero.step(), in that case the correct gradient tensor indeed needs to be returned.

To clarify this, the `ZeroRedundancyOptimizer.step()` called before buckets are rebuilt do not actually do any optimization, so given what @rohan-varma explained, I think actually we do not need to return the bucket tensor -- just a tensor of the same dimension. I misspoke in the previous comment.